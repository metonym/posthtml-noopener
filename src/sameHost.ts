export const sameHost = (href: string) => /^(\.|\/)/.test(href);
