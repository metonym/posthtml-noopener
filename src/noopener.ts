import { PostHTMLTree } from 'posthtml';
import { sameHost } from './sameHost';

function noopener() {
  return function plugin(tree: PostHTMLTree) {
    tree.match(
      {
        tag: 'a',
        attrs: { target: '_blank', href: new RegExp(/\S+/) }
      },
      node => {
        const { rel, href } = node.attrs!;
        const invalidRel = rel === undefined || !rel.trim();
        const notSameHost = !sameHost(href.trim());

        if (invalidRel && notSameHost) {
          node.attrs!.rel = 'noopener noreferrer';
        }

        return node;
      }
    );
  };
}

export { noopener };
