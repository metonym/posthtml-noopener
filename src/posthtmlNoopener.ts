import { sameHost } from './sameHost';

function posthtmlNoopener() {
  return function plugin(tree: ITree) {
    tree.match(
      {
        tag: 'a',
        attrs: {
          href: new RegExp(/\S+/),
          target: '_blank'
        }
      },
      node => {
        const { rel, href } = (node as INodeTargetBlank).attrs;

        const invalidRel = rel === undefined || !rel.trim();
        const notSameHost = !sameHost(href.trim());

        if (invalidRel && notSameHost) {
          (node as INodeTargetBlank).attrs.rel = 'noopener noreferrer';
        }

        return node;
      }
    );
  };
}

export { posthtmlNoopener };

export interface INode {
  tag?: string;
  attrs?: {
    id?: string;
    class?: string;
    target?: string;
    href?: string;
    rel?: string;
  };
  content?: INode[];
}

interface INodeTargetBlank extends INode {
  attrs: {
    target: '_blank';
    href: string;
    rel?: string;
  };
}

type Match = string | RegExp;

interface IMatcher {
  tag?: string;
  attrs?: {
    id?: Match;
    class?: Match;
    target?: Match;
    href?: Match;
    rel?: Match;
  };
  content?: IMatcher[];
}

interface ITree {
  match: (matcher: IMatcher | IMatcher[], node: (node: INode) => INode) => void;
}
