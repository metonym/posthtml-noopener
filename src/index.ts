import { noopener } from './noopener';

export default noopener;
export { noopener, noopener as posthtmlNoopener };
