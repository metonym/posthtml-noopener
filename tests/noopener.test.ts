import posthtml from "posthtml";
import { describe, test, expect } from "vitest";
import { noopener } from "../src";

interface ITestCase {
  name: string;
  input: string;
}

const testCases: ITestCase[] = [
  {
    name: "Link without `rel` attribute",
    input: `<a href="http://" target="_blank"></a>`,
  },
  {
    name: "Link with empty `href` attribute is ignored",
    input: `<a href="" target="_blank"></a>`,
  },
  {
    name: "Link with empty `rel` attribute",
    input: `<a href="http://" target="_blank" rel=""></a>`,
  },
  {
    name: "Link to same host is ignored",
    input: `<a href="/" target="_blank"></a>`,
  },
];

describe("posthtmlNoopener", () => {
  testCases.forEach((testCase) => {
    test(`'${testCase.name}' matches the snapshot`, () => {
      posthtml()
        .use(noopener())
        .process(testCase.input)
        .then(({ html }) => {
          expect(html).toMatchSnapshot();
        });
    });
  });
});
