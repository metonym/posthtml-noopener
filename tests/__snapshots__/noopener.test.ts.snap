// Vitest Snapshot v1

exports[`posthtmlNoopener > 'Link to same host is ignored' matches the snapshot 1`] = `"<a href=\\"/\\" target=\\"_blank\\"></a>"`;

exports[`posthtmlNoopener > 'Link with empty \`href\` attribute is ignored' matches the snapshot 1`] = `"<a href=\\"\\" target=\\"_blank\\"></a>"`;

exports[`posthtmlNoopener > 'Link with empty \`rel\` attribute' matches the snapshot 1`] = `"<a href=\\"http://\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\"></a>"`;

exports[`posthtmlNoopener > 'Link without \`rel\` attribute' matches the snapshot 1`] = `"<a href=\\"http://\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\"></a>"`;
